nemo/collections/tts/models/spectrogram_enhancer.py

```diff
@@ -0,0 +1,382 @@
+
+        # generate noise
+        if zs is not None and ws is not None:
+            raise ValueError(
+                "Please specify either zs or ws or neither, but not both. It is not clear which one to use."
+            )
+
+        if zs is None:
+            zs = self.generate_zs(batch_size, mixing)
+        if ws is None:
+            ws = [self.G.style_mapping(z) for z in zs]
+        if noise is None:
+            noise = self.generate_noise(batch_size)
+
+        # normalize if needed, mask and pad appropriately
+        if normalize:
+            condition = self.normalize_spectrograms(condition)
+        condition = self.pad_spectrogram(condition)
+        condition = mask(condition, lengths)
+
+        # the main call
+        enhanced_spectrograms = self.G(condition, lengths, ws, noise)
+
+        # denormalize if needed, mask and remove padding
+        if normalize:
+            enhanced_spectrograms = self.unnormalize_spectrograms(enhanced_spectrograms)
+        enhanced_spectrograms = mask(enhanced_spectrograms, lengths)
+        enhanced_spectrograms = enhanced_spectrograms[:, :, :, :max_length]
+
+        return enhanced_spectrograms
+
+    def prepare_batch(self, batch):
+        if not isinstance(self.spectrogram_model, FastPitchModel):
+            raise NotImplementedError(
+                f"{type(self.spectrogram_model)} is not supported, please implement batch preparation for this model."
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
+        mels, mel_lens = self.spectrogram_model.preprocessor(input_signal=audio, length=audio_lens)
+
+        mels_pred, *_, pitch = self.spectrogram_model(
+            text=text,
+            durs=durs,
+            pitch=pitch,
+            speaker=speaker,
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
+            attn_prior=attn_prior,
+            mel_lens=mel_lens,
+            input_lens=text_lens,
+        )
+
+        mels = mels.cpu()
+        mel_lens = mel_lens.cpu()
+        mels_pred = mels_pred.cpu()
+
+        batch_size, *_ = mels.shape
+
+        targets = []
+        conditions = []
+        lengths = []
+
+        for i in range(batch_size):
+            l = mel_lens[i].item()
+            lengths.append(l)
+
+            targets.append(mels[i, :, :l].T)
+            conditions.append(mels_pred[i, :, :l].T)
+
+        target = nn.utils.rnn.pad_sequence(targets, batch_first=True)
+        condition = nn.utils.rnn.pad_sequence(conditions, batch_first=True)
+        lengths = torch.LongTensor(lengths).unsqueeze(-1)
+
+        target = rearrange(target, "b l c -> b 1 c l")
+        condition = rearrange(condition, "b l c -> b 1 c l")
+
+        return self.move_to_correct_device((target, condition, lengths))
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
```

Comment:
You should add a corresponding `validation_step` as well.

Reply:
I'm not convinced validation step is necessary here - all of the losses are rather uninformative in deciding which model is "better"

Reply:
Losses in GAN training are always difficult to interpret, but we should track the losses in wandb if only for debugging and making sure things do not become NaN and such.

Reply:
Will do after implementing preprocessing

Reply:
Current decision is not to